diffkemp/simpll/InstPatternComparator.cpp

```diff
@@ -0,0 +1,655 @@
+    // structure names.
+    if (int Res = cmpTypes(TyL, TyR)) {
+        if (!namesMatch(TyL->getStructName(), TyR->getStructName(), IsLeftSide))
+            return Res;
+    }
+
+    if (int Res = cmpNumbers(GEPL->getNumOperands(), GEPR->getNumOperands()))
+        return Res;
+
+    for (unsigned i = 0, e = GEPL->getNumOperands(); i != e; ++i) {
+        if (int Res = cmpValues(GEPL->getOperand(i), GEPR->getOperand(i)))
+            return Res;
+    }
+
+    return 0;
+}
+
+/// Compare a module function instruction with a pattern instruction along
+/// with their operands.
+/// Note: Parts of this function have been adapted from FunctionComparator.
+/// Therefore, LLVM licensing also applies here. See the LICENSE information
+/// in the appropriate llvm-lib subdirectory for more details.
+int InstPatternComparator::cmpOperationsWithOperands(
+        const Instruction *L, const Instruction *R) const {
+    bool needToCmpOperands = true;
+
+    // Clear newly mapped holders.
+    NewlyMappedValuesL.clear();
+    NewlyMappedValuesR.clear();
+    NewlyMappedInputL.clear();
+    NewlyMappedInputR.clear();
+
+    // Compare the instruction and its operands.
+    if (int Res = cmpOperations(&*L, &*R, needToCmpOperands))
+        return Res;
+    if (needToCmpOperands) {
+        assert(L->getNumOperands() == R->getNumOperands());
+
+        for (unsigned i = 0, e = L->getNumOperands(); i != e; ++i) {
+            Value *OpL = L->getOperand(i);
+            Value *OpR = R->getOperand(i);
+
+            if (int Res = cmpValues(OpL, OpR))
+                return Res;
+            assert(cmpTypes(OpL->getType(), OpR->getType()) == 0);
+        }
+    }
+    // Map the instructions to each other.
+    InstMatchMap[R] = L;
+    return 0;
+}
+
+/// Compare a module function basic block with a pattern basic block.
+int InstPatternComparator::cmpBasicBlocks(const BasicBlock *BBL,
+                                          const BasicBlock *BBR) const {
+    auto InstL = BBL->begin(), InstLE = BBL->end();
+    auto InstR = BBR->begin(), InstRE = BBR->end();
+    auto TermL = BBL->getTerminator(), TermR = BBR->getTerminator();
+
+    // Jump to the currently compared instruction pair.
+    jumpToInst(InstL, PositionL);
+    jumpToInst(InstR, PositionR);
+
+    while (InstL != InstLE && InstR != InstRE) {
+        PositionL = &*InstL;
+        PositionR = &*InstR;
+
+        // Check whether the compared pattern ends at this instruction.
+        auto InstRMetadata = ParentPattern->MetadataMap.find(&*InstR);
+        if (InstRMetadata != ParentPattern->MetadataMap.end()
+            && InstRMetadata->second.PatternEnd)
+            return 0;
+
+        // If at the end of only one basic block, leave the rest to
+        // unconditionally connected basic blocks (if there are any).
+        if (&*InstL == TermL && &*InstR != TermR
+            && TermL->getNumSuccessors() == 1) {
+            return 1;
```

Comment:
Because in the first case (i.e., when `1` gets returned), the pattern block (the block containing `InstR`) is not fully matched. Hence, a match "failure" is returned. This can be corrected (and the comparison continued) by the `InstPatternComparator::matchPattern` if a single successor of the module block exists.

In other words, when a module block ends before a pattern block is fully matched, the matching of basic blocks fails. However, if the following module block is connected unconditionally, the primary algorithm will continue matching, starting with the first instruction in the successor module block and the next instruction from the original pattern block.

Practically, the opposite is true in the second case. There, the pattern block gets fully matched, so the matching of the blocks succeeds (and hence, `0` is returned). Again, the matching may continue if further basic blocks (this time on the pattern side) exist.

I hope the code makes more sense now!